File: solutions/python/code-carl/ch04_string/t2_lc0541_reverse_str.py
"""LeetCode 541. Reverse String II
https://leetcode.cn/problems/reverse-string-ii/
"""


class Solution:
    def reverseStr(self, s: str, k: int) -> str:
        chars = list(s)
        for i in range(0, len(s), 2 * k):
            chars[i:i + k] = chars[i:i + k][::-1]
        return "".join(chars)


if __name__ == "__main__":
    solution = Solution()
    assert solution.reverseStr("abcdefg", 2) == "bacdfeg"
    assert solution.reverseStr("abcd", 2) == "bacd"
    print("All tests passed.")

File: solutions/code-carl/ch04_string/T2_LC0541_ReverseStr.java
/**
 * Practice template for:
 * - LeetCode 541. Reverse String II
 *   https://leetcode.cn/problems/reverse-string-ii/
 */
public class T2_LC0541_ReverseStr {
    public String reverseStr(String s, int k) {
        char[] chars = s.toCharArray();
        int begin;
        for (begin = 0; begin < chars.length; begin += 2 * k) {
            reverse(chars, begin, Math.min(begin + k - 1, chars.length - 1));
        }

        return new String(chars);
    }

    public void reverse(char[] s, int l, int r) {
        while (l < r) {
            char tmp = s[l];
            s[l] = s[r];
            s[r] = tmp;
            l++;
            r--;
        }
    }
}
